Forwarder error on HTTP2 --> HTTP headers

Hello, 
If the oxy forwarder is used on HTTP2 protocol environments, the redirection does not work on Safari on Mac Os or IOs. The bug has been discovered using Traefik Proxy (see the link).

https://github.com/containous/traefik/issues/307#issuecomment-220918326

And seems to be related to the fact that the protocol version is manually overridden to http 1.1 , https://github.com/vulcand/oxy/blob/master/forward/fwd.go#L220

Any hints on which could be the best way to solve this?

I ran into this bug using Traefik proxying in front of Nginx.
1. Safari makes a http2 request to Traefik.
2. Traefik makes a http1.1 request to Nginx.
3. The response from Nginx contains "Connection: keep-alive" which gets copied to the http2 response sent to Safari.

Safari goes banas and tries to get the page repeatedly a lot of times. Which the user sees as responsetimes measured in seconds. Eventually safari gives up and shows you the page.
Chrome handles it more gracefully.

Curl show the error in plain text
```bash
curl -v --http2 https://mysite.com

* TCP_NODELAY set
* Copying HTTP/2 data in stream buffer to connection buffer after upgrade: len=0
* Using Stream ID: 1 (easy handle 0x7fa57b806200)
> GET / HTTP/1.1
> Host: mysite.com
> User-Agent: curl/7.50.1
> Accept: */*
>
* Connection state changed (MAX_CONCURRENT_STREAMS updated)!
* http2 error: Invalid HTTP header field was received: frame type: 1, stream: 1, name: [connection], value: [keep-alive]
* HTTP/2 stream 1 was not closed cleanly: PROTOCOL_ERROR (err 1)
* Connection #0 to host mysite.com left intact
curl: (92) HTTP/2 stream 1 was not closed cleanly: PROTOCOL_ERROR (err 1)
```

If I disable keep-alive in Nginx `keepalive_timeout 0s;` it works like a charm.

So I guess the `connection` header should be removed by Oxy when proxying from http1.1 to http2.

I have tcpdump logs if that helps.

--- a/forward/fwd.go
+++ b/forward/fwd.go
@@ -13,11 +13,12 @@ import (
 	"time"
 
 	"crypto/tls"
-	log "github.com/Sirupsen/logrus"
-	"github.com/vulcand/oxy/utils"
 	"net"
 	"net/http/httputil"
 	"reflect"
+
+	log "github.com/sirupsen/logrus"
+	"github.com/vulcand/oxy/utils"
 )
 
 // ReqRewriter can alter request headers and body
@@ -205,6 +206,16 @@ func (f *httpForwarder) serveBufferedHTTP(w http.ResponseWriter, req *http.Reque
 		log.Infof("vulcand/oxy/forward/httpbuffer: Round trip: %v, code: %v, duration: %v",
 			req.URL, response.StatusCode, time.Now().UTC().Sub(start))
 	}
+
+	// Connection: references headers that should be treated as hop by hop
+	if c := response.Header.Get("Connection"); c != "" {
+		for _, f := range strings.Split(c, ",") {
+			if f = strings.TrimSpace(f); f != "" {
+				response.Header.Del(f)
+			}
+		}
+	}
+	utils.RemoveHeaders(response.Header, HopHeaders...)
 
 	utils.CopyHeaders(w.Header(), response.Header)
 	w.WriteHeader(response.StatusCode)

